scalyr_agent/copying_manager/worker.py

```diff
@@ -964,31 +975,33 @@ def handle_completed_callback(result):
             # TODO:  This might not be bullet proof here.  We copy __log_processors and then update it at the end
             # We could be susceptible to exceptions thrown in the middle of this method, though now should.
 
-            # Copy the processor list because we may need to remove some processors if they are done.
-            processor_list = self.__log_processors[:]
-            self.__log_processors = []
-
             add_events_request.close()
 
             total_bytes_copied = 0
 
-            for processor in processor_list:
-                # Iterate over all the processors, seeing if we had a callback for that particular processor.
-                if processor.unique_id in all_callbacks:
-                    # noinspection PyCallingNonCallable
-                    # If we did have a callback for that processor, report the status and see if we callback is done.
-                    (closed_processor, bytes_copied) = all_callbacks[
-                        processor.unique_id
-                    ](result)
-                    keep_it = not closed_processor
-                    total_bytes_copied += bytes_copied
-                else:
-                    keep_it = True
-
-                if keep_it:
-                    self.__log_processors.append(processor)
-                else:
-                    processor.close()
+            with self._closed_files_checkpoints_lock:
```

Comment:
I worry that we are holding onto this lock for too long.  Can any of the calls you are making while this lock is held block?  Is there a way to structure this so that the lock can be held for a much shorter time?  For example, while we are going through this loop, store which the checkpoints for the closed processors in a temporary variable.  Once out of the loop, copy it back into the `_closed_files_checkpoints` while you hold the lock.

Reply:
Oh, sure. Yeah, there's no need to hold the lock during the whole loop, thanks.